### PR TITLE
Marea processor

### PR DESCRIPTION
**Pull request by marea-cobb**
Updated statistics plugin to include minimum, maximum, range, mode, sum, skewness, kurtosis, count, and trimean.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+sudo: required
+language: Go
+go:
+        - 1.5.3
```
Comment (candysmurf) [inline .travis.yml]: Should we use go1.7 and go1.6.3?

File: statistics/statistics.go [hunk 2]
```diff
@@ -41,6 +42,7 @@ type Plugin struct {
 	bufferMaxSize int
 	bufferCurSize int
 	bufferIndex   int
+	logger        *log.Logger
```
Comment (candysmurf) [inline statistics/statistics.go]: Are you sure that you like to include logger inside the statistics?

File: statistics/statistics.go [hunk 3]
```diff
@@ -53,26 +55,31 @@ func Meta() *plugin.PluginMeta {
...
-		bufferIndex:   0}
+		bufferIndex:   0,
+		logger:        log.New()}
 	return p
 }
 
-func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[string]float64, error) {
-	result := make(map[string]float64)
+// calculateStats calaculates the descriptive statistics for buff
+func (p *Plugin) calculateStats(buff interface{}) (map[string]interface{}, error) {
```
Comment (candysmurf) [inline statistics/statistics.go]: why did you change the map value from float64 to interface?

Comment (marea-cobb) [inline statistics/statistics.go]: The functions range and mode return a slice of float6.

File: statistics/statistics.go [hunk 4]
```diff
@@ -135,30 +144,120 @@ func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[strin
...
 	result["99%-ile"] = val
 
+	val, err = stats.Min(buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	result["Kurtosis"] = p.Kurtosis(buffer)
+
+	result["Minimum"] = val
```
Comment (candysmurf) [inline statistics/statistics.go]: shouldn't this line be closer to your mix calculation?

File: statistics/statistics.go [hunk 5]
```diff
@@ -135,30 +144,120 @@ func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[strin
...
+}
+
+// concatNameSpace combines an array of namespces into a single string
 func concatNameSpace(namespace []string) string {
 	completeNamespace := ""
-	for i := 0; i < len(namespace); i++ {
-		completeNamespace += namespace[i]
-
+	for _, ns := range namespace {
+		completeNamespace += ns
```
Comment (candysmurf) [inline statistics/statistics.go]: Using append is better. Also, there is no delimiter between namespace elements?

File: statistics/statistics.go (hunk 4 above)
Comment (marea-cobb) [inline statistics/statistics.go]: Yes, good catch. it should be moved up to line 151.

File: statistics/statistics.go [hunk 6]
```diff
@@ -187,10 +287,11 @@ func (p *Plugin) GetConfigPolicy() (*cpolicy.ConfigPolicy, error) {
 	return cp, nil
 }
 
+// Process processes the data, inputs the data into this' buffer and calls the descriptive statistics method
 func (p *Plugin) Process(contentType string, content []byte, config map[string]ctypes.ConfigValue) (string, []byte, error) {
 
-	logger := log.New()
-	logger.Println("Statistics Processor started")
+	//logger := log.New()
```
Comment (candysmurf) [inline statistics/statistics.go]: remove commented out code

File: statistics/statistics.go [hunk 7]
```diff
@@ -207,25 +308,34 @@ func (p *Plugin) Process(contentType string, content []byte, config map[string]c
...
-		logger.Printf("Data received %v", metrics[i].Data())
-		p.insertInToBuffer(metrics[i].Data(), metrics[i].Namespace().Strings())
-	}
+	for _, metric := range metrics {
+		p.logger.Printf("Data received %v", metric.Data())
+		p.insertInToBuffer(metric.Data(), metric.Namespace().Strings())
+		p.updateCounters()
 
-	p.updateCounters()
+		//p.updateCounters()
```
Comment (candysmurf) [inline statistics/statistics.go]: remove

File: statistics/statistics.go [hunk 8]
```diff
@@ -207,25 +308,34 @@ func (p *Plugin) Process(contentType string, content []byte, config map[string]c
...
 
-	for i, _ := range metrics {
+		//	for _, metric := range metrics {
```
Comment (candysmurf) [inline statistics/statistics.go]: remove

File: statistics/statistics.go (hunk 2 above)
Comment (marea-cobb) [inline statistics/statistics.go]: This was recommended from a review Balaji forwarded me from Izabella Raulin, Katarzyna Zabrocka, and Kasia.

File: statistics/statistics.go [hunk 9]
```diff
@@ -207,25 +308,34 @@ func (p *Plugin) Process(contentType string, content []byte, config map[string]c
...
+		var err error
 		if p.bufferCurSize < p.bufferMaxSize {
-			metrics[i].Data_, _ = p.calculateStats(p.buffer[concatNameSpace(metrics[i].Namespace().Strings())][0:p.bufferCurSize], logger)
+			metric.Data_, err = p.calculateStats(p.buffer[concatNameSpace(metric.Namespace().Strings())][0:p.bufferCurSize])
+			if err != nil {
+				return "", nil, err
+			}
 		} else {
```
Comment (candysmurf) [inline statistics/statistics.go]: don't need else as returned on an error

File: statistics/statistics.go (hunk 3 above)
Comment (candysmurf) [inline statistics/statistics.go]: using map[string][]float64 is better than map[string]interface as it only allows float64.

File: statistics/statistics_test.go [hunk 10]
```diff
@@ -92,10 +92,9 @@ func TestStatisticsProcessorMetrics(t *testing.T) {
 				time.Sleep(3)
```
Comment (candysmurf) [inline statistics/statistics_test.go]: why do you have to sleep here?

File: statistics/statistics_test.go [hunk 11]
```diff
@@ -105,25 +104,19 @@ func TestStatisticsProcessorMetrics(t *testing.T) {
...
-			_, receivedData, _ := movingAverageObj.Process("snap.gob", buf.Bytes(), nil)
-
-			var metricsNew []plugin.PluginMetricType
+			movingAverageObj.Process("snap.gob", buf.Bytes(), nil)
 
-			//Decodes the content into MetricType
-			dec := gob.NewDecoder(bytes.NewBuffer(receivedData))
-			dec.Decode(&metricsNew)
+			var metricsNew []plugin.MetricType
 			So(metrics, ShouldNotResemble, metricsNew)
```
Comment (candysmurf) [inline statistics/statistics_test.go]: Do you have any definite number to verify the calculation?

File: statistics/statistics_test.go [hunk 12]
```diff
@@ -135,13 +128,12 @@ func TestStatisticsProcessorMetrics(t *testing.T) {
 
 			_, receivedData, _ := movingAverageObj.Process("snap.gob", buf.Bytes(), nil)
 
-			var metricsNew []plugin.PluginMetricType
+			var metricsNew []plugin.MetricType
 
 			//Decodes the content into MetricType
 			dec := gob.NewDecoder(bytes.NewBuffer(receivedData))
 			dec.Decode(&metricsNew)
 			So(metrics, ShouldNotResemble, metricsNew)
```
Comment (candysmurf) [inline statistics/statistics_test.go]: Same as the above. It didn't really verify the calculation.

File: statistics/statistics_test.go (hunk 11 above)
Comment (marea-cobb) [inline statistics/statistics_test.go]: Rashmi and I were wondering this. The test is using a random number generator each time so we don't have a means to actually verify the value. We thought this seemed odd but wasn't sure if this was "best practice" or if there was an end-to-end test that focused on the actual values.

Comment (candysmurf) [inline statistics/statistics_test.go]: if we can have mock tests to verify the actual vs expected, that would be awesome.

File: statistics/statistics.go [hunk 13]
```diff
@@ -62,13 +65,15 @@ func New() *Plugin {
 	return p
 }
 
-func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[string]float64, error) {
-	result := make(map[string]float64)
+// calculateStats calaculates the descriptive statistics for buff
+func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[string][]float64, error) {
```
Comment (candysmurf) [inline statistics/statistics.go]: don't pass log.Logger as the method argument. It does not need to be passed around. You can directly use it.
Such as log.Logger.Printf(...)

File: statistics/statistics.go [hunk 14]
```diff
@@ -93,72 +98,161 @@ func (p *Plugin) calculateStats(buff interface{}, logger *log.Logger) (map[strin
...
+		num += math.Pow(val-mean, 4)
+		den += math.Pow(val-mean, 2)
+	}
+	return []float64{float64(len(buffer)) * num / math.Pow(den, 2)}
 }
 
-func (p *Plugin) insertInToBuffer(val interface{}, ns []string) {
+// concatNameSpace combines an array of namespces into a single string
+func concatNameSpace(namespace []string) string {
+	completeNamespace := strings.Join(namespace, ", ")
```
Comment (candysmurf) [inline statistics/statistics.go]: probably use slash as the delimiter "/".

**Comment by candysmurf** [comment]
@marea-cobb, if you can remove 2 binary files(main, snap-plugin-processor-statistics) and remove the // build Linux from file main.go. We may get this PR merged. As we chatted, you'll work on concrete tests to verify calculation next.

**Comment by candysmurf** [comment]
LGTM
